apps/aecore/lib/aecore/tx/data_tx.ex

```diff
@@ -49,66 +49,109 @@ defmodule Aecore.Tx.DataTx do
-      end
+    tx_type_state = get_tx_type_state(chainstate, tx.type)
+
+    case tx.payload
+         |> tx.type.init()
+         |> tx.type.process_chainstate(
+           tx.sender,
+           tx.fee,
+           tx.nonce,
+           block_height,
+           accounts_state_tree,
+           tx_type_state
+         ) do
+      {:error, reason} ->
+        {:error, reason}
+
+      {new_accounts_state_tree, new_tx_type_state} ->
+        new_chainstate =
+          if tx.type == SpendTx do
+            chainstate
+          else
+            Map.put(chainstate, tx.type.get_chain_state_name(), new_tx_type_state)
+          end
+
+        {:ok, Map.put(new_chainstate, :accounts, new_accounts_state_tree)}
+    end
+  end
+
+  @doc """
+  Gets the given transaction type state,
+  if there is any stored in the chainstate
+  """
+  @spec get_tx_type_state(ChainState.chainstate(), atom()) :: map()
+  def get_tx_type_state(chainstate, tx_type) do
+    type = tx_type.get_chain_state_name()
+    Map.get(chainstate, type, %{})
+  end
+
+  @spec validate_sender(Wallet.pubkey(), ChainState.chainstate()) :: :ok | {:error, String.t()}
+  def validate_sender(sender, %{accounts: account}) do
+    case AccountStateTree.get(account, sender) do
+      {:ok, _account_key} ->
+        :ok
 
-    if !nonce_valid?(accounts_state_tree, tx) do
-      throw({:error, "Nonce is too small"})
+      :none ->
+        {:error, "#{__MODULE__}: The senders key: #{inspect(sender)} doesn't exist"}
     end
+  end
 
-    {new_accounts_state_tree, new_tx_type_state} =
-      tx.payload
-      |> tx.type.init()
-      |> tx.type.process_chainstate!(
-        tx.sender,
-        tx.fee,
-        tx.nonce,
-        block_height,
-        accounts_state_tree,
-        tx_type_state
-      )
-
-    new_chainstate =
-      if tx.type == SpendTx do
-        chainstate
-      else
-        Map.put(chainstate, tx.type.get_chain_state_name(), new_tx_type_state)
-      end
-
-    Map.put(new_chainstate, :accounts, new_accounts_state_tree)
+  @spec validate_nonce(ChainState.account(), DataTx.t()) :: :ok | {:error, String.t()}
+  def validate_nonce(accounts_state, tx) do
+    if tx.nonce > Account.nonce(accounts_state, tx.sender) do
+      :ok
+    else
+      {:error, "#{__MODULE__}: Nonce is too small: #{inspect(tx.nonce)}"}
+    end
   end
 
-  @spec nonce_valid?(ChainState.accounts(), DataTx.t()) :: boolean()
-  def nonce_valid?(accounts_state, tx) do
-    tx.nonce > Account.nonce(accounts_state, tx.sender)
+  @spec preprocess_check(DataTx.t(), ChainState.chain_state(), non_neg_integer()) ::
+          :ok | {:error, String.t()}
+  def preprocess_check(
+        %DataTx{
+          type: type,
+          payload: payload,
+          sender: sender,
+          fee: fee,
+          nonce: nonce
+        } = tx,
+        %{accounts: accounts} = chainstate,
+        block_height
+      ) do
+    sender_account_state = Account.get_account_state(accounts, sender)
```

Comment:
I would suggest invoking `validate_nonce` and `validate_sender` inside this function instead of in `Aecore.Chain.Chainstate`

Reply:
They are in the ChainState module for readability reasons. That way the code has good flow of reading, and understandability. If they are inside the DataTx it's not visible that they are called.

Reply:
If they are inside DataTx, DataTx is more self contained, ChainState "knows less about what DataTx does" and for me personally it has better flow of reading.